src/python/pants/engine/rules.py

```diff
@@ -41,12 +44,36 @@
 PANTS_RULES_MODULE_KEY = "__pants_rules__"
 
 
+def implicitly(*args) -> dict[str, Any]:
```

Comment:
I'm just staring at this function trying to figure out what it means and I got nothing :sweat_smile:

Reply:
This is just the existing concept where the actual arguments to a `@rule` are computed "automatically"/implicitly from the given argument and context. This syntax makes that existing concept explicit, while still allowing for positional and keyword arguments to be passed (not implemented yet in this PR). https://github.com/pantsbuild/pants/discussions/18905#discussioncomment-5821244

The naming comes approximately from Scala's "implicit parameters", aka "contextual parameters": https://docs.scala-lang.org/tour/implicit-parameters.html

I'm very open to new names for this function. Will open another discussion point on #18905.